### PR TITLE
Upgrade ethash and use ethash error codes

### PR DESCRIPTION
**Pull request by chfast**
Upstream change: https://github.com/chfast/ethash/pull/187.

Here I have some concerns about readability because it is not directly visible that `verify()` returns error code.

### CODE REVIEW

**Comment by chfast** [comment]
> I like this but I wonder if we should wait till ethash lib changes are merged into an official release.

Yes. I just wanted to see how it looks here.

File: cmd/check_pow.cpp [hunk 1]
```diff
@@ -131,15 +131,16 @@ int main(int argc, char* argv[]) {
...
+            const auto sealh256{ethash::hash256_from_bytes(seal_hash.bytes)};
+            const auto mixh256{ethash::hash256_from_bytes(header->mix_hash.bytes)};
+            if (const auto ec = ethash::verify(*epoch_context, sealh256, mixh256, nonce, boundary256); ec) {
                 auto result{ethash::hash(*epoch_context, sealh256, nonce)};
                 auto b{to_bytes32({boundary256.bytes, 32})};
                 auto f{to_bytes32({result.final_hash.bytes, 32})};
                 auto m{to_bytes32({result.mix_hash.bytes, 32})};
 
                 std::cout << "\n Pow Verification error on block " << block_num << " : \n"
+                          << "Error: " << ec << "\n"
```
Comment (chfast) [inline cmd/check_pow.cpp:143]: I'm not sure where this is used. But here you can either print final or mix hash, depending on error code.

File: cmd/check_pow.cpp [hunk 2]
```diff
@@ -131,15 +131,16 @@ int main(int argc, char* argv[]) {
 
             auto boundary256{header->boundary()};
             auto seal_hash(header->hash(/*for_sealing =*/true));
-            ethash::hash256 sealh256{*reinterpret_cast<ethash::hash256*>(seal_hash.bytes)};
-            ethash::hash256 mixh256{*reinterpret_cast<ethash::hash256*>(header->mix_hash.bytes)};
-            if (!ethash::verify(*epoch_context, sealh256, mixh256, nonce, boundary256)) {
+            const auto sealh256{ethash::hash256_from_bytes(seal_hash.bytes)};
```
[the remark this answers is not in the document]
Comment (AndreaLanfranchi) [reply cmd/check_pow.cpp:134]: Yeah right

File: core/silkworm/consensus/ethash/engine.cpp [hunk 3]
```diff
@@ -56,7 +56,7 @@ ValidationResult ConsensusEngineEthash::validate_seal(const BlockHeader& header)
     std::memcpy(mixh256.bytes, header.mix_hash.bytes, 32);
 
     uint64_t nonce{endian::load_big_u64(header.nonce.data())};
-    return ethash::verify(*epoch_context, sealh256, mixh256, nonce, boundary256) ? ValidationResult::kOk
-                                                                                 : ValidationResult::kInvalidSeal;
+    return ethash::verify(*epoch_context, sealh256, mixh256, nonce, boundary256) ? ValidationResult::kInvalidSeal
```
[the remark this answers is not in the document]
Comment (AndreaLanfranchi) [reply core/silkworm/consensus/ethash/engine.cpp]: Made the change

File: cmd/check_pow.cpp (hunk 1 above)
Comment (AndreaLanfranchi) [reply cmd/check_pow.cpp:143]: Not a big deal. `check_pow` is only a debug tool

**Comment by AndreaLanfranchi** [comment]
Would be super nice if we could also bundle all of this with https://github.com/chfast/ethash/issues/185

**Review by AndreaLanfranchi — APPROVED** [review]

**Pull request merged**